README.md
<a href="https://www.gotoiot.com/">
    <img src="doc/gotoiot-logo.png" alt="logo" title="Goto IoT" align="right" width="60" height="60" />
</a>

# Web App Full Stack Base

Este proyecto es una aplicación web fullstack que se ejecuta sobre el ecosistema `Docker`. Está compuesta por un compilador de `TypeScript` que te permite utilizar este superset de JavaScript para poder programar un `cliente web`. También tiene un servicio en `NodeJS` que te permite ejecutar código en backend y al mismo tiempo disponibilizar el código del cliente web para interactar con el servicio.

La aplicación IoT de base que viene con este proyecto se encarga de crear una tabla llamada `Devices` en la base de datos, y la idea es que vos puedas desarrollar el código de backend y frontend que te permita controlar desde el navegador el estado de los devices de un hogar inteligente - _como pueden ser luces, TVs, ventiladores, persianas, enchufes y otros_ - y almacenar los estados de cada uno en la base de datos.

En esta imagen podés ver una posible implementación del cliente web que controla los artefactos del hogar.

![architecture](doc/client-app.png)

## Comenzando 🚀

Esta sección es una guía con los pasos escenciales para que puedas poner en marcha la aplicación.

<details><summary><b>Mira los pasos necesarios</b></summary><br>

### Instalar las dependencias

Para correr este proyecto es necesario que instales `Docker` y `Docker Compose`.

En [este artículo](https://www.gotoiot.com/pages/articles/docker_installation_linux/) publicado en la web están los detalles para instalar Docker y Docker Compose en una máquina Linux.

En caso que quieras instalar las herramientas en otra plataforma o tengas algún incoveniente, podes leer la documentación oficial de [Docker](https://docs.docker.com/get-docker/) y también la de [Docker Compose](https://docs.docker.com/compose/install/).

Continua con la descarga del código cuando tengas las dependencias instaladas y funcionando.

### Descargar el código

Para descargar el código, utiliza este comando desde la terminal o con tu IDE favorito.

```
git clone https://github.com/NGSalvo/app-fullstack-base.git
```

> No es necesario contar con una cuenta en Github.

### Ejecutar la aplicación

Para ejecutar la aplicación tenes que correr el comando `docker-compose up` desde la raíz del proyecto. Este comando va a descargar las imágenes de Docker de node, de typescript, de la base datos y del admin de la DB, y luego ponerlas en funcionamiento.

Para acceder al cliente web ingresa a a la URL [http://localhost:8000/](http://localhost:8000/).

Si pudiste acceder al cliente web significa que la aplicación se encuentra corriendo bien.

> Si te aparece un error la primera vez que corres la app, detené el proceso y volvé a iniciarla. Esto es debido a que el backend espera que la DB esté creada al iniciar, y en la primera ejecución puede no alcanzar a crearse. A partir de la segunda vez el problema queda solucionado.

</details>

Continuá explorando el proyecto una vez que lo tengas funcionando.

## Configuraciones de funcionamiento 🔩

Al crearse la aplicación se ejecutan los contenedores de Docker de cada servicio, se crea la base de datos y sus tablas. A continuación podés encontrar info si querés cambiar la estructura de la DB o bien sus configuraciones de acceso.

## Detalles principales 🔍

En esta sección vas a encontrar las características más relevantes del proyecto.

<details><summary><b>Mira los detalles más importantes de la aplicación</b></summary><br>
<br>

### Arquitectura de la aplicación

Como ya pudiste ver, la aplicación se ejecuta sobre el ecosistema Docker, y en esta imagen podés ver el diagrama de arquitectura.

![architecture](doc/architecture.png)

### El cliente web

El cliente web es una Single Page Application que se comunica con el servicio en NodeJS mediante JSON a través de requests HTTP. Puede consultar el estado de dispositivos en la base de datos (por medio del servicio en NodeJS) y también cambiar el estado de los mismos. Los estilos del código están basados en [**Material Design**](https://materializecss.com/).

### El servicio web

El servicio en **NodeJS** posee distintos endpoints para comunicarse con el cliente web mediante requests HTTP enviando **JSON** en cada transacción. Procesando estos requests es capaz de comunicarse con la base de datos para consultar y controlar el estado de los dispositivos, y devolverle una respuesta al cliente web también en formato JSON. Así mismo el servicio es capaz de servir el código del cliente web.

### El compilador de TypeScript

**TypeScript** es un lenguaje de programación libre y de código abierto desarrollado y mantenido por Microsoft. Es un superconjunto de JavaScript, que esencialmente añade tipos estáticos y objetos basados en clases. Para esta aplicación se usa un compilador de TypeScript basado en una imagen de [Harmish](https://hub.docker.com/r/harmish) en Dockerhub, y está configurado para monitorear en tiempo real los cambios que se realizan sobre el directorio **src/frontend/ts** y automáticamente generar código compilado a JavaScript en el directorio **src/frontend/js**. Los mensajes del compilador aparecen automáticamente en la terminal al ejecutar el comando **docker-compose up**.

### Ejecución de servicios

Los servicios de la aplicación se ejecutan sobre **contenedores de Docker**, así se pueden desplegar de igual manera en diferentes plataformas. Los detalles sobre cómo funcionan los servicios los podés ver directamente en el archivo **docker-compose.yml**.

### Organización del proyecto

En la siguiente ilustración podés ver cómo está organizado el proyecto para que tengas en claro qué cosas hay en cada lugar.

```sh
├── db                          # directorio de la DB
│   ├── data                    # estructura y datos de la DB
│   └── dumps                   # directorio de estructuras de la DB
│       └── smart_home.sql      # estructura con la base de datos "smart_home"
├── doc                         # documentacion general del proyecto
└── src                         # directorio codigo fuente
│   ├── backend                 # directorio para el backend de la aplicacion
│   │   ├── index.js            # codigo principal del backend
│   │   ├── mysql-connector.js  # codigo de conexion a la base de datos
│   │   ├── package.json        # configuracion de proyecto NodeJS
│   │   └── package-lock.json   # configuracion de proyecto NodeJS
│   └── frontend                # directorio para el frontend de la aplicacion
│       ├── js                  # codigo javascript que se compila automáticamente
│       ├── static              # donde alojan archivos de estilos, imagenes, fuentes, etc.
│       ├── ts                  # donde se encuentra el codigo TypeScript a desarrollar
│       └── index.html          # archivo principal del cliente HTML
├── docker-compose.yml          # archivo donde se aloja la configuracion completa
├── README.md                   # este archivo
├── CHANGELOG.md                # archivo para guardar los cambios del proyecto
├── LICENSE.md                  # licencia del proyecto
```

</details>

## Detalles de implementación 💻

En esta sección podés ver los detalles específicos de funcionamiento del código y que son los siguientes.

<details><summary><b>Mira los detalles de implementación</b></summary><br>

### Agregar un dispositivo

Para agregar un dispositvo se debe presionar sobre el botón que se muestra en la **Figura 1**.

![Agregar nuevo](doc/device-abm/00-add.png)

<figcaption><p align=center>Figura 1.</p></figcaption>

Llenar los datos y presionar el botón **Guardar** como se muestra en la **Figura 2**.

![Guardar nuevo](doc/device-abm/01-modal_empty.png)

<figcaption><p align=center>Figura 2.</p></figcaption>

Una vez guardado debe desplegarse la notificación de guardado exitoso. Ver **Figura 3**.

![Notificación creación](doc/device-abm/03-notificacion.png)

<figcaption><p align=center>Figura 3.</p></figcaption>

### Editar un dispositivo

Para editar, presionar sobre el botón **Editar** en uno de los dispositivos como se muestra en la **Figura 4**.

![Editar](doc/device-abm/04-edit.png)

<figcaption><p align=center>Figura 4.</p></figcaption>

Cuando se abre el modal, traerá los datos del dispositivo seleccionado. Editar los campos que se deseen y presionár el botón **Guardar** como se muestra en la **Figura 5**.

![Guardar editado](doc/device-abm/05-editar.png)

<figcaption><p align=center>Figura 5.</p></figcaption>

Una vez guardado debe desplegarse la notificación de cambio exitoso. Ver **Figura 6**.

![Notificación edición](doc/device-abm/06-notificacion_edit.png)

<figcaption><p align=center>Figura 6.</p></figcaption>

### Eliminar un dispositivo

Para eliminar un dispositivo, presionar sobre el botón **Eliminar** en uno de los dispositivos como se muestra en la **Figura 7**.

![Eliminar](doc/device-abm/07-delete.png)

<figcaption><p align=center>Figura 7.</p></figcaption>

En el modal, presionar sobre el botón **Sí** para confirmar la eliminación como se muestra en la **Figura 8**.

![Confirmar eliminado](doc/device-abm/08-confirmar_eliminar.png)

<figcaption><p align=center>Figura 8.</p></figcaption>

Una vez confirmado debe desplegarse la notificación de eliminación exitosa. Ver **Figura 9**.

![Notificación eliminación](doc/device-abm/09-notificacion_eliminar.png)

<figcaption><p align=center>Figura 9.</p></figcaption>

### Frontend

El frontend está divido en 3 grandes clases:

- main.ts
- MyFramework.ts
- ViewMainPage.ts

`main.ts` está compuesto por la logica principal del programa. En ella se encuentran el manejo de eventos de la UI.

`MyFramework.ts` contiene algunas utilidades para obtención de partes del DOM. También posee la estructura principal de los llamados AJAX.

`ViewMainPage.ts` contiene solo modificaciones del DOM, como los agregados de los dispositivos. También cuenta con la función de renderización y refresco de los dispositivos.

Interactua con el backend por medio de las llamadas AJAX, realizando las correspondientes peticions GET (para obtener información del servidor), PUT (para actualizar un dispositivo), POST (para agregar un dispositivo) y DELETE (para eliminación de un dispositivo).

### Backend

El backend consiste de todos los endpoints para que el cliente web interactue. Estos endpoints contienen la lógica para escribir y obtener información de un archivo `devices.json`.

Los métodos utilizados son GET, POST, PUT y DELETE.

GET obtiene información.
POST agrega nueva información.
PUT modifica información.
DELETE elimina información.

Para fines demostrativos se utilizaron los 4 tipos, pero dado que el proyecto fue realizado con archivos, hubiese alcanzado simplemente con POST y GET.

<details><summary><b>Ver los endpoints disponibles</b></summary><br>

## Obtener todos los dispositivos

```json
{
  "endpoint": "/devices",
  "method": "get",
  "request_headers": "application/json",
  "request_body": "",
  "response_code": 200,
  "response_body": {
    "devices": [
      {
        "id": 1,
        "name": "Lámpara 1",
        "description": "Luz Living",
        "state": 0,
        "type": 0
      }
    ]
  }
}
```

## Obtener un dispositivo

```json
{
  "endpoint": "/devices/:id",
  "method": "get",
  "request_headers": "application/json",
  "request_body": "",
  "response_code": 200,
  "response_body": {
    "devices": {
      "id": 1,
      "name": "Lámpara 1",
      "description": "Luz Living",
      "state": 0,
      "type": 0
    }
  }
}
```

## Actualizar un dispositivo

```json
{
  "endpoint": "/devices/:id",
  "method": "put",
  "request_headers": "application/json",
  "request_body": "Cualquier parámetro de un dispositivo",
  "response_code": 200,
  "response_body": { "message": "Se realizó el cambio correctamente" }
  }
}
```

## Crear un dispositivo

```json
{
  "endpoint": "/devices/create",
  "method": "post",
  "request_headers": "application/json",
  "request_body": "",
  "response_code": 201,
  "response_body": {
    "devices": { "message": "Se creó el dispositivo satisfactoriamente" }
  }
}
```

## Eliminar un dispositivo

```json
{
  "endpoint": "/devices/:id",
  "method": "delete",
  "request_headers": "application/json",
  "request_body": "Cualquier parámetro de un dispositivo",
  "response_code": 200,
  "response_body": { "message": "Se eliminó el dispositivo Lámpara 1" }
  }
}
```

</details>

</details>

## Tecnologías utilizadas 🛠️

En esta sección podés ver las tecnologías más importantes utilizadas.

<details><summary><b>Mira la lista completa de tecnologías</b></summary><br>

- [Docker](https://www.docker.com/) - Ecosistema que permite la ejecución de contenedores de software.
- [Docker Compose](https://docs.docker.com/compose/) - Herramienta que permite administrar múltiples contenedores de Docker.
- [Node JS](https://nodejs.org/es/) - Motor de ejecución de código JavaScript en backend.
- [Material Design](https://material.io/design) - Bibliotecas de estilo responsive para aplicaciones web.
- [TypeScript](https://www.typescriptlang.org/) - Superset de JavaScript tipado y con clases.

</details>

## Sobre Goto IoT 📖

Goto IoT es una plataforma que publica material y proyectos de código abierto bien documentados junto a una comunidad libre que colabora y promueve el conocimiento sobre IoT entre sus miembros. Acá podés ver los links más importantes:

- **[Sitio web](https://www.gotoiot.com/):** Donde se publican los artículos y proyectos sobre IoT.
- **[Github de Goto IoT:](https://github.com/gotoiot)** Donde están alojados los proyectos para descargar y utilizar.
- **[Comunidad de Goto IoT:](https://groups.google.com/g/gotoiot)** Donde los miembros de la comunidad intercambian información e ideas, realizan consultas, solucionan problemas y comparten novedades.
- **[Twitter de Goto IoT:](https://twitter.com/gotoiot)** Donde se publican las novedades del sitio y temas relacionados con IoT.
- **[Wiki de Goto IoT:](https://github.com/gotoiot/doc/wiki)** Donde hay información de desarrollo complementaria para ampliar el contexto.

## Autores 👥

Las colaboraciones principales fueron realizadas por:

- **[Agustin Bassi](https://github.com/agustinBassi)**: Ideación, puesta en marcha y mantenimiento del proyecto.
- **[Ernesto Giggliotti](https://github.com/ernesto-g)**: Creación inicial del frontend, elección de Material Design.
- **[Brian Ducca](https://github.com/brianducca)**: Ayuda para conectar el backend a la base de datos, puesta a punto de imagen de Docker.
- **[Nicolás Salvo](https://github.com/NGSalvo)**: realizó el ejercicio para la materia DAW de la Carrera de Especialización en Internet de las Cosas.

También podés mirar todas las personas que han participado en la [lista completa de contribuyentes](https://github.com/###/contributors).

## Licencia 📄

Este proyecto está bajo Licencia ([MIT](https://choosealicense.com/licenses/mit/)). Podés ver el archivo [LICENSE.md](LICENSE.md) para más detalles sobre el uso de este material.

---

**Copyright © Goto IoT 2021** ⌨️ [**Website**](https://www.gotoiot.com) ⌨️ [**Group**](https://groups.google.com/g/gotoiot) ⌨️ [**Github**](https://www.github.com/gotoiot) ⌨️ [**Twitter**](https://www.twitter.com/gotoiot) ⌨️ [**Wiki**](https://github.com/gotoiot/doc/wiki)

---

Se utilizó como recurso gratuito:

- <a target="_blank" href="https://icons8.com/icon/yGKsUh4h-5W2/light-dimmer">Light Dimmer</a> icon by <a target="_blank" href="https://icons8.com">Icons8</a>
- <a target="_blank" href="https://icons8.com/icon/XBJfETMfZHpS/light">Light</a> icon by <a target="_blank" href="https://icons8.com">Icons8</a>


CHANGELOG.md
# DAW Base App - Changes Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 3.0.0 -

- Project modification
  - Adds device's CRUD (Backend)
  - Adds device's CRUD (Frontend)

## 2.2.0

- Project modification
  - Adds TypeScript compiler service to Docker Compose
  - Reestructures frontend folder for TypeScript
  - Adds new info to README accordingly
  - Changes project architecture image

## 2.1.0

- Project modification
  - Enhaces README accordingly to Goto IoT
  - Adds example of finished application
  - Removes unnecessary frontend images
  - Changes src code folders names

## 2.0.0

- Project modification
  - Changes project and organization names
  - Removes Typescript container
  - Removes Typescript Code
  - Executes Javascript code directly
  - Changes licence to MIT
  - Modifies README accordingly

## 1.0.0

- Project creation
  - Docker Compose implementation for whole project.
  - Typescript compilation into docker-compose.
  - MySQL 5.7 DB Server.
  - PHPMyAdmin.
  - NodeJS backend application.
  - Materialize CSS framework.


src/frontend/ts/main.ts
type DeviceInt = {
  id?: string;
  name: string;
  description: string;
  state: number;
  type: number;
};

// Estados del formulario
type FormStatus = 'editing' | 'creating' | 'deleting' | 'none';

class Main
  implements
    EventListenerObject,
    GETResponseListener,
    POSTResponseListener,
    DELETEResponseListener,
    PUTResponseListener
{
  public myFramework: MyFramework;
  public viewMainPage: ViewMainPage;
  public formStatus: FormStatus = 'none';
  public currentDevice: DeviceInt = this.defaultDevice();
  public modalInstances;
  public selectInstances;
  private URI = 'http://localhost:8000/devices';
  public devices: Array<DeviceInt>;
  public elementListener = new WeakMap();
  constructor() {}

  main(): void {
    this.myFramework = new MyFramework();

    this.myFramework.requestGET(this.URI, this);

    this.viewMainPage = new ViewMainPage();

    const modals = document.querySelectorAll('.modal');
    this.modalInstances = M.Modal.init(modals, { dismissible: false });

    const comboboxes = document.querySelectorAll('select');
    this.selectInstances = M.FormSelect.init(comboboxes);
    comboboxes[0].addEventListener('change', this);
  }

  // Manejo de los eventos
  handleEvent(event: Event): void {
    // console.log(event.target);

    let element: HTMLElement = this.myFramework.getElementByEvent(event);

    // Evento del estado del dispositivo
    if (element.id.includes('status_')) {
      let type = element.getAttribute('type') === 'range' ? '1' : '0';
      let value: number | boolean;
      let state: number;
      if (type === '1') {
        value = this.viewMainPage.getRangeValueById(element.id);
      } else {
        value = this.viewMainPage.getSwitchStateById(element.id) ? 1 : 0;
      }
      state = value;
      let id = element.id.slice(7);
      console.log(state);
      this.myFramework.requestPUT(`${this.URI}/${id}`, { state }, this);
      this.currentDevice = this.devices.find((device) => device.id === id);
      this.currentDevice.state = state;
      this.updateDevice(id, this.currentDevice);
      console.log(this.devices);
      return;
    }

    // Evento del boton de editar
    if (element.classList.contains('edit_dev')) {
      // console.log('EDIT:', this.currentDevice);
      this.formStatus = 'editing';
      document.getElementById('modal1').querySelector('h4').textContent =
        'Editar dispositivo';
      let id = element.id.slice(5);
      this.myFramework.requestGET(`${this.URI}/${id}`, this);
      setTimeout(() => {
        this.setFormData();
        let type = this.currentDevice.type;
        this.viewMainPage.showModalComponentType(type, this.currentDevice);
      }, 200);
      return;
    }

    // Evento del boton de eliminar
    if (element.classList.contains('delete_dev')) {
      this.formStatus = 'deleting';
      let id = element.id.slice(7);
      this.currentDevice.id = id;
      return;
    }

    // Evento del boton de guardar (MODAL)
    if (element.classList.contains('btn-save')) {
      // Evento del boton guardar (MODAL) si esta editando
      if (this.formStatus === 'editing') {
        this.currentDevice = this.getFormData();
        let id = this.currentDevice.id;
        this.myFramework.requestPUT(
          `${this.URI}/${id}`,
          this.currentDevice,
          this,
        );
        this.updateDevice(id, this.currentDevice);
        this.clearForm();
      }
      // Evento del boton guardar (MODAL) si esta creando
      if (this.formStatus === 'creating') {
        this.currentDevice = this.getFormData();
        // console.log(this.currentDevice);
        this.myFramework.requestPOST(
          `${this.URI}/create`,
          this.currentDevice,
          this,
        );
        this.addNewDevice(this.currentDevice);
        this.clearForm();
      }

      // Evento del boton guardar (MODAL) si esta eliminando
      if (this.formStatus === 'deleting') {
        let id = this.currentDevice.id;
        this.myFramework.requestDELETE(`${this.URI}/${id}`, this);
        this.removeDevice(id);
        this.clearForm();
      }
      // Se renderiza los elementos creados, modificados u eliminados
      this.getDevices();
      return;
    }

    if (element.classList.contains('btn-cancel')) {
      this.clearForm();
      return;
    }

    if (element.textContent === 'add') {
      this.formStatus = 'creating';
      document.getElementById('modal1').querySelector('h4').textContent =
        'Crear dispositivo';
      return;
    }

    if (element.id === 'deviceType') {
      let type = +(
        this.myFramework.getElementById('deviceType') as HTMLSelectElement
      ).value;
      this.viewMainPage.showModalComponentType(type, this.currentDevice);
    }
  }

  // -------Manejo de AJAX-------
  handleGETResponse(status: number, response: any): void {
    let res = JSON.parse(response);
    if (status === 200) {
      let data: DeviceInt[] | DeviceInt = JSON.parse(response);
      if (Array.isArray(data)) {
        this.devices = data;
        this.viewMainPage.showDevices(data);

        let element: HTMLElement;
        data.forEach((data) => {
          element = this.myFramework.getElementById('dev_' + data.id);
          element.addEventListener('click', this);
        });
      } else {
        this.currentDevice = data;
      }
    }
    if (status === 404) {
      M.toast({ html: `${res.message}`, classes: 'toast-failure' });
    } else {
      console.log(`Error: ${status}`);
    }
  }

  handlePOSTResponse(status: number, response: any): void {
    let res = JSON.parse(response);
    if (status === 200 || status === 201) {
      M.toast({ html: `${res.message}`, classes: 'toast-success' });
    }
    if (status === 400) {
      M.toast({ html: `${res.message}`, classes: 'toast-failure' });
    } else {
      console.log(`Error: ${status}`);
    }
  }

  handleDELETEResponse(status: number, response: any): void {
    let res = JSON.parse(response);
    if (status === 200) {
      M.toast({ html: `${res.message}`, classes: 'toast-success' });
    }
    if (status === 404) {
      M.toast({ html: `${res.message}`, classes: 'toast-failure' });
    } else {
      console.log(`Error: ${status}`);
    }
  }

  handlePUTResponse(status: number, response: any): void {
    let res = JSON.parse(response);
    if (status === 200) {
      M.toast({ html: `${res.message}`, classes: 'toast-success' });
    }
    if (status === 404) {
      M.toast({ html: `${res.message}`, classes: 'toast-failure' });
    } else {
      console.log(`Error: ${status}`);
    }
  }
  // -------Fin manejo de AJAX-------

  // Devuelve un dispositivo con valores por defecto
  defaultDevice(): DeviceInt {
    return {
      name: '',
      description: '',
      state: 0,
      type: 0,
    };
  }

  // -------Manejo de formulario-------
  // Deja el formulario como estaba en un principio
  clearForm() {
    this.currentDevice = this.defaultDevice();
    this.setFormData();
    this.formStatus = 'none';
  }

  // Obtengo todos los datos del modal
  getFormData(): DeviceInt {
    let id = this.currentDevice.id;
    let name = (
      this.myFramework.getElementById('deviceName') as HTMLInputElement
    ).value;
    let description = (
      this.myFramework.getElementById('deviceDescription') as HTMLInputElement
    ).value;
    let type = +(
      this.myFramework.getElementById('deviceType') as HTMLSelectElement
    ).value;
    let state: number;
    if (type === 1) {
      state = state = (
        this.myFramework.getElementById('deviceState') as HTMLInputElement
      ).valueAsNumber;
    } else {
      state = (
        this.myFramework.getElementById('deviceState') as HTMLInputElement
      ).checked
        ? 1
        : 0;
    }

    return { id, name, description, state, type };
  }

  // Configuro todos los datos del modal
  setFormData() {
    // console.log('setFormData', this.currentDevice);
    (this.myFramework.getElementById('deviceName') as HTMLInputElement).value =
      this.currentDevice.name;
    (
      this.myFramework.getElementById('deviceDescription') as HTMLInputElement
    ).value = this.currentDevice.description;
    if (this.currentDevice.type === 1) {
      (this.myFramework.getElementById('deviceState') as HTMLInputElement)
        .valueAsNumber;
    } else {
      (
        this.myFramework.getElementById('deviceState') as HTMLInputElement
      ).checked = this.currentDevice.state === 0 ? false : true;
    }
    (
      this.myFramework.getElementById('deviceType') as HTMLSelectElement
    ).options[this.currentDevice.type].selected = true;
    this.selectInstances = M.FormSelect.init(
      this.myFramework.getElementById('deviceType'),
    );
  }
  // -------Fin manejo de formulario-------

  // -------Manejo de dispositivos en memoria-------
  addNewDevice(device: DeviceInt) {
    this.devices = [...this.devices, device];
  }
  removeDevice(id: string) {
    this.devices = this.devices.filter((device) => device.id !== id);
  }
  updateDevice(id: string, update: DeviceInt) {
    const deviceIndex = this.devices.findIndex((device) => id === device.id);

    if (deviceIndex > -1) {
      this.devices[deviceIndex] = {
        ...this.devices[deviceIndex],
        ...update,
      };
    } else {
      console.log('Error al actualizar. No lo encuentra');
    }
  }
  getDevices() {
    // Se vuelven a renderizar los elementos
    this.viewMainPage.showDevices(this.devices);

    // Se agregan los listener a los nuevos elementos
    let element: HTMLElement;
    this.devices.forEach((device) => {
      element = this.myFramework.getElementById('dev_' + device.id);
      this.elementListener.set(
        element,
        element.addEventListener('click', this),
      );
    });
  }
  // -------Fin manejo de dispositivos en memoria-------
}

window.addEventListener('load', function () {
  const main = new Main();
  main.main();

  // Agregar listener desde otro lugar para demostrar que se puede
  let modal1: HTMLElement = main.myFramework.getElementById('modal1');
  modal1.addEventListener('click', main);
  let modal2: HTMLElement = main.myFramework.getElementById('modal2');
  modal2.addEventListener('click', main);
  let btnCreateDevice: HTMLElement =
    main.myFramework.getElementById('btnCreateDevice');
  btnCreateDevice.addEventListener('click', main);
  // main.viewMainPage.showModalComponentType(main.currentDevice);
});


src/frontend/ts/MyFramework.ts
interface GETResponseListener {
  handleGETResponse(status: number, response: any): void;
}

interface POSTResponseListener {
  handlePOSTResponse(status: number, response: any): void;
}

interface DELETEResponseListener {
  handleDELETEResponse(status: number, response: any): void;
}

interface PUTResponseListener {
  handlePUTResponse(status: number, response: any): void;
}
class MyFramework {
  constructor() {}

  getElementById(elementId: string): HTMLElement {
    return document.getElementById(elementId);
  }

  getElementByEvent(event: Event): HTMLElement {
    return event.target as HTMLElement;
  }

  // Devuelve el elemento padre
  getElementParent(element: HTMLElement): HTMLElement {
    return element.parentElement;
  }

  // Elimina elementos hijos
  removeChildren(element: HTMLElement): void {
    while (element.lastChild) {
      element.removeChild(element.lastChild);
    }
  }

  requestGET(url: string, listener: GETResponseListener) {
    let xhr: XMLHttpRequest;
    xhr = new XMLHttpRequest();
    xhr.onreadystatechange = () => {
      if (xhr.readyState === 4) {
        if (xhr.status === 200) {
          listener.handleGETResponse(xhr.status, xhr.response);
        } else {
          listener.handleGETResponse(xhr.status, null);
        }
      }
    };
    xhr.open('GET', url, true);
    xhr.send(null);
  }

  requestPOST(url: string, data: object, listener: POSTResponseListener) {
    let xhr: XMLHttpRequest;
    xhr = new XMLHttpRequest();
    xhr.onreadystatechange = () => {
      if (xhr.readyState === 4) {
        if (xhr.status === 200 || xhr.status === 201) {
          listener.handlePOSTResponse(xhr.status, xhr.response);
        } else {
          listener.handlePOSTResponse(xhr.status, null);
        }
      }
    };
    xhr.open('POST', url);
    xhr.setRequestHeader('Content-Type', 'application/json;charset=UTF-8');
    xhr.send(JSON.stringify(data));
  }

  requestDELETE(url: string, listener: DELETEResponseListener) {
    let xhr: XMLHttpRequest;
    xhr = new XMLHttpRequest();
    xhr.onreadystatechange = () => {
      if (xhr.readyState === 4) {
        if (xhr.status === 200) {
          listener.handleDELETEResponse(xhr.status, xhr.response);
        } else {
          listener.handleDELETEResponse(xhr.status, null);
        }
      }
    };
    xhr.open('DELETE', url, true);
    xhr.send(null);
  }

  requestPUT(url: string, data: object, listener: PUTResponseListener) {
    let xhr: XMLHttpRequest;
    xhr = new XMLHttpRequest();
    xhr.onreadystatechange = () => {
      if (xhr.readyState === 4) {
        if (xhr.status === 200) {
          listener.handlePUTResponse(xhr.status, xhr.response);
        } else {
          listener.handlePUTResponse(xhr.status, null);
        }
      }
    };
    xhr.open('PUT', url);
    xhr.setRequestHeader('Content-Type', 'application/json;charset=UTF-8');
    xhr.send(JSON.stringify(data));
  }
}


src/backend/index.js
//=======[ Settings, Imports & Data ]==========================================

const PORT = 3000;

const express = require('express');
const app = express();
// const utils = require('./mysql-connector');

const fs = require('fs');
const path = require('path');
const { v4: uuidv4 } = require('uuid');

let devices = require('./devices.json');
const devicesPath = path.join(__dirname, 'devices.json');

// to parse application/json
app.use(express.json());
// to serve static files
app.use(express.static('/home/node/app/static/'));

//=======[ Main module code ]==================================================

// Endpoint: /devices
// Retorna json de dispositivos
app.get('/devices/', function (req, res, next) {
  res.status(200).json(devices);
});

// Endpoint: /devices/:id
// id - el ID del dispositivo
// Retorna el json de un dispositivo
app.get('/devices/:id', function (req, res, next) {
  const id = req.params.id;
  const deviceIndex = getDeviceIndexById(id);
  if (deviceIndex > -1) {
    res.status(200).json(devices[deviceIndex]);
  } else {
    res
      .status(404)
      .send({ message: `No se encontró ningún dispositivo con id:${id}` });
  }
});

// Endpoint: /devices/:id
// id - el ID del dispositivo
// Actualiza el dispositivo
app.put('/devices/:id', function (req, res, next) {
  const id = req.params.id;
  const deviceIndex = getDeviceIndexById(id);

  if (deviceIndex > -1) {
    devices[deviceIndex] = {
      ...devices[deviceIndex],
      ...req.body,
    };
    // writeToFile(devices, devicesPath);
    res.status(200).send({ message: `Se realizó el cambio correctamente` });
  } else {
    res
      .status(404)
      .send({ message: `No se encontró ningún dispositivo con id:${id}` });
  }
});

// Endpoint: /devices/create
// Crea un nuevo dispositivo
app.post('/devices/create', function (req, res, next) {
  const { name, description, state, type } = req.body;

  if (
    !isDefined(name) ||
    !isDefined(description) ||
    !isDefined(state) ||
    !isDefined(type)
  ) {
    res.status(400).json({ message: 'Falta un parámetro en la consulta' });
  }

  const id = getNewId();
  const newDevice = {
    id,
    name,
    description,
    state,
    type,
  };
  devices.push(newDevice);
  // writeToFile(devices, devicesPath);
  res
    .status(201)
    .send({ message: 'Se creó el dispositivo satisfactoriamente' });
});

// Endpoint: /devices/:id
// id - el ID del dispositivo
// Elimina un dispositivo
app.delete('/devices/:id', function (req, res, next) {
  const id = req.params.id;
  const deviceIndex = getDeviceIndexById(id);
  if (deviceIndex > -1) {
    const device = devices.splice(deviceIndex, 1);
    // writeToFile(devices, devicesPath);
    res
      .status(200)
      .send({ message: `Se eliminó el dispositivo ${device[0].name}` });
  } else {
    res
      .status(404)
      .send({ message: `No se encontró un dispositivo con id:${id}` });
  }
});

app.listen(PORT, function (req, res) {
  console.log('NodeJS API running correctly');
});

// id - id del dispositivo que se busca
// Retorna el indice del dispositivo
function getDeviceIndexById(id) {
  return devices.findIndex((device) => id === device.id);
}

// file - el archivo a escribir
// path - la direccion del archivo a escribir
// Escribe/Sobreescribe un archivo
function writeToFile(file, path) {
  const fileStringified = JSON.stringify(file);
  fs.writeFile(path, fileStringified, (err) => {
    if (err) {
      console.log(err);
    }
  });
}

// item - elemento a comparar
// Devuelve verdadero si el item esta definido, sino devuelve falso.
function isDefined(item) {
  if (item === undefined) {
    return false;
  }
  return true;
}

// Devuelve un ID unico
function getNewId() {
  return uuidv4();
}

//=======[ End of file ]=======================================================
